### PR TITLE
ecds: add dynamic config provider usage in downstream network filters

### LINKED ISSUE #14696: ecds: add support for network filters
Implement ECDS for network filters in listeners. API added [here](https://github.com/envoyproxy/envoy/pull/14674)

### PR DESCRIPTION
**Pull request by ohadvano**
Additional Description: Follow up to #28407, this PR adds dynamic config provider usage in downstream network filters, with relevant integration tests. A follow up PR will add support for upstream network filters. Link: #14696.
Risk Level: low (guarded by config)
Testing: Integration tests
Docs Changes: ECDS API doc, network filter, ECDS config doc
Release Notes: None
Platform Specific Features: None

### CODE REVIEW

**Comment by ohadvano** [comment]
/assign @kyessenov 
I'll appreciate your review on this as a follow-up. The actual source change is minimal, this is mostly integration tests.

**Comment by ohadvano** [comment]
/unassign @kyessenov 
Seems like Kuat is out for vacation.
/assign @htuch 
Could you please review this as a follow up to the previous PR?
Thanks

**Comment by htuch** [comment]
I'm also out for the next 1.5 weeks. @adisuissa @mattklein123 do you think you folks could take this one? Thanks.

**Comment by ohadvano** [comment]
/unassign @lizan
/unassign @htuch 
/assign @adisuissa 
/assign @mattklein123 
Assigning to you as you could hopefully review this. Thanks

**Comment by ohadvano** [comment]
Ping @mattklein123, @adisuissa

**Comment by mattklein123** [comment]
Sorry I don't have time to review this in detail so will wait for @adisuissa to do a first pass, thanks.

File: api/envoy/service/extension/v3/config_discovery.proto
Comment (adisuissa) [inline api/envoy/service/extension/v3/config_discovery.proto:21]: High level comment:
Can you point me to Envoy's support for upstream network filters?
If it is not supported, I suggest removing the unused-code/comments as they may cause more confusion.

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/network_extension_discovery_integration_test.cc:252]: nit: I think these can be constexpr and maybe string_view

File: docs/root/configuration/overview/extension.rst
Comment (adisuissa) [inline docs/root/configuration/overview/extension.rst:56]: Please clarify when the new network filter configuration will take into effect.
IIUC this will only be for new connections, and won't be applied to ongoing (long-living) connections.

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/network_extension_discovery_integration_test.cc:297]:
```suggestion
  // Send 1st config update with TTL 1s, and have network filter drain 5 bytes of data.
```

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/network_extension_discovery_integration_test.cc:211]: nit:
```suggestion
    const std::string expected_data = data_.substr(bytes_drained);
```

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/network_extension_discovery_integration_test.cc:371]: Is there a way to monitor connections that try to use an invalid configuration?

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/network_extension_discovery_integration_test.cc:521]:
```suggestion
// Validate that a network filter update should fail if the subscribed extension configuration make filter
```

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/network_extension_discovery_integration_test.cc:525]: FWIW, I don't think that listener and network filters should be terminal, but I guess this is "inherited" from http filters.

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/network_extension_discovery_integration_test.cc:706]: 2 high-level questions:
1. Are there planned integration tests that test the interaction between listener updates (LDS) and network-filter updates (ECDS)?
2. Can the HCM be updated? If so I think more tests are needed in that space because of the lifetime of some objects.
3. Please add tests that have ongoing connections during an ECDS update and show that the "old" config is still used for these filters.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this! Looks very good!
Mostly nits and some high-level comment about further testing.

cc @yanjunxiang-google as a person who worked on this before.

File: test/integration/network_extension_discovery_integration_test.cc
Comment (ohadvano) [reply test/integration/network_extension_discovery_integration_test.cc:525]: ``tcp_proxy``/``hcm`` for example are defined as terminal, i.e can only be used as last network filter

File: test/integration/network_extension_discovery_integration_test.cc
Comment (ohadvano) [reply test/integration/network_extension_discovery_integration_test.cc:371]: I am not sure there currently is.
I believe we would want a metric similar to listener's ``extension_config_missing``:
https://github.com/envoyproxy/envoy/blob/d2c5c0c9341e16f939f8e85da15850bb2970e244/source/extensions/listener_managers/listener_manager/listener_impl.cc#L958

This stat is increased when a connection is closed due to missing configuration. I intend to add an equivalent to network filters after this one is merged.

File: docs/root/configuration/overview/extension.rst
Comment (ohadvano) [reply docs/root/configuration/overview/extension.rst:56]: done, thanks

File: test/integration/network_extension_discovery_integration_test.cc
Comment (ohadvano) [reply test/integration/network_extension_discovery_integration_test.cc]: It could but then it would require creating a string in functions that take std::string which repeats quick a few.

File: test/integration/network_extension_discovery_integration_test.cc
Comment (ohadvano) [reply test/integration/network_extension_discovery_integration_test.cc:800]: I added LDS integration and a test that verifies that old config is used for connections that were established before ECDS update. Due to (3), I believe the behavior here would be the same, connections will get the HCM config that is updated at the time of connection establishment. Did you have another behavior in mind which I missed?

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [reply test/integration/network_extension_discovery_integration_test.cc:429]: Ack. That's desired, as otherwise it will be hard to monitor such cases.

File: test/integration/network_extension_discovery_integration_test.cc
Comment (adisuissa) [reply test/integration/network_extension_discovery_integration_test.cc:580]: Ack.
I guess the use of the term [is_terminal here](https://github.com/envoyproxy/envoy/pull/28477/files#diff-791ac42e27b3c608b85afdc77097fe7a1e99b6e590be8fbbce60f5f165c91c4fR88) is somewhat misleading.

**Review by adisuissa — DISMISSED** [review]
Thanks, LGTM.
/assign-from @envoyproxy/senior-maintainers

File: test/integration/network_extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [inline test/integration/network_extension_discovery_integration_test.cc:518]: I think it is valuable to add integration test cases for pure static filters as this:

https://github.com/envoyproxy/envoy/blob/06e3e7d59fddc6541cecf399a8adc64092df7ed8/test/integration/listener_extension_discovery_integration_test.cc#L496

Comment (ohadvano) [reply test/integration/network_extension_discovery_integration_test.cc:518]: @yanjunxiang-google can you please elaborate on the value that you see? Is it not just like not using ECDS at all (which I believe is covered in regular network filter tests)?

File: api/envoy/service/extension/v3/config_discovery.proto
Comment (ohadvano) [reply api/envoy/service/extension/v3/config_discovery.proto:21]: I see. I added this in advance in order to split PRs logic instead of having PRs with big source change. I'll add upstream network filter support soon so it won't be stale much longer.

File: test/integration/network_extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [inline test/integration/network_extension_discovery_integration_test.cc:606]: Thanks for config dump test.

It's probably valuable to add a network ecds filter config dump test here:

https://github.com/envoyproxy/envoy/blob/06e3e7d59fddc6541cecf399a8adc64092df7ed8/test/server/admin/config_dump_handler_test.cc#L809

Comment (ohadvano) [reply test/integration/network_extension_discovery_integration_test.cc:606]: Thanks for the insight, added.

**Comment by ohadvano** [comment]
@mattklein123 seems lke Adi approved, could you please take a look?

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
